providers/asrockrack/firmware.go

```diff
@@ -0,0 +1,231 @@
+package asrockrack
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"os"
+	"strconv"
+	"strings"
+
+	"github.com/pkg/errors"
+
+	"github.com/bmc-toolbox/bmclib/devices"
+	bmclibErrs "github.com/bmc-toolbox/bmclib/errors"
+)
+
+
+
+// FirmwareInstall uploads and initiates firmware update for the component
+func (a *ASRockRack) FirmwareInstall(ctx context.Context, component, applyAt string, forceInstall bool, reader io.Reader) (jobID string, err error) {
+	var size int64
+	if file, ok := reader.(*os.File); ok {
```

Comment:
If the reader is not an `os.File`, is that a problem? Why is this necessary?

Reply:
This was to determine the file size to pass as the `Content-Length` header in the multipart upload,

https://github.com/bmc-toolbox/bmclib/blob/inventory-firmware/providers/asrockrack/helpers.go#L210

At some point, I need to take a look at if the ASRR API service still requires the content length.

Reply:
If length is required, maybe consider a different interface like [`fs.File`](https://pkg.go.dev/io/fs#File)  or you could declare an interface like 
```go
type LengthReader interface {  
    io.Reader
    Len() int
}
```

which is satisfied by `bytes.Buffer` and `strings.Reader` and provide a helper for `os.File/fs.File` like

```go
type lr struct {
  fs.File
}

func (lr *r) Len() int {
    info, err := r.Stat()
    if err != nil {
        panic(err)
    }
    return int(info.Size())
}

// LengthReaderFromFile returns a LengthReader from an fs.File
func LengthReaderFromFile(f fs.File) LengthReader {
    return &lr{f}
}
```